Add deprecated dao banner

AquiGorka opened the pull request:
These changes include:

- A modal manager using React context (components down the tree will be able to trigger modal popups now)
- A banner that is shown when the DAO is deprecated (default false)
- The banner can trigger a modal when the user clicks on the `more info` button

Banner:
![image](https://user-images.githubusercontent.com/3072458/46412352-99ea1000-c71e-11e8-9480-33e234de7d4d.png)

Modal:
![image](https://user-images.githubusercontent.com/3072458/46412362-a53d3b80-c71e-11e8-8ec8-12014f2960f7.png)

In action:
![deprecated-modal](https://user-images.githubusercontent.com/3072458/46412840-ec77fc00-c71f-11e8-9c04-5541e6b7bb4c.gif)

sohkai commented on src/components/Modal/Modal.js:
```diff
@@ -0,0 +1,78 @@
+import { Button, Text, breakpoint } from '@aragon/ui'
+
+const WrapModal = styled(EscapeOutside)`
+  height: calc(100vh - 2em);
```
We should definitely give you a `small` breakpoint to avoid this :smile:

<img src="https://user-images.githubusercontent.com/4166642/46490404-34724e00-c808-11e8-8162-b3229a5bf4c0.png" height=400>

AquiGorka replied:
Screens smaller than `768px`? I don't think it is needed, most probably those sizes belong to smartphones.

bpierre replied:
The reason why there is no `small` breakpoint is to ensure the base CSS is working with any size of screen (mobile first) with exceptions (breakpoints) for the larger sizes, rather than the other way. https://github.com/aragon/aragon-ui/blob/master/src/utils/styles/breakpoint.js#L7

sohkai commented on src/Wrapper.js:
```diff
@@ -166,10 +172,14 @@ class Wrapper extends React.Component {
-      <React.Fragment>
+      <ModalProvider>
+        <ModalView />
+        {isDaoDeprecated && <DeprecatedBanner dao={dao} />}
```
Thoughts on moving both the modal and the deprecation banner up one level, to be in the [App](https://github.com/aragon/aragon/blob/master/src/App.js)? Right now the banner shows up during onboarding but it's not really obvious because it's here.

AquiGorka replied:
I'll adhere to both your thoughts on this. @sohkai and @bpierre 👆

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+        old-app.aragon.org/
+        {dao}
+      </StyledSafeLink>{' '}
+      for a few months after 1/11/18.
```
Edits to this copy:

```
On Nov. 1, 2018, this DAO will no longer be accessible on app.aragon.one. To help migrate, and
in case you still need access to this DAO, we will continue hosting this version of Aragon Core
on old-app.aragon.org/#/{dao} until the end of March, 2019.
```

We could use `<time>`s for the dates, and I'd recommend keeping it in text form, to avoid confusion with different orders of parsing (in North America `1/11/2018` would be read as Jan. 11, 2018).

I also made up [March, 2019](https://github.com/aragon/governance/pull/34/files#r222691994) but I think it's better to give a precise date here.

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+      accessible at the current URL. To help manage this migration, and in case
+      you still need access to your old DAOs, we’ll be hosting this DAO created
+      with 0.5 at{' '}
+      <StyledSafeLink href="https://old-app.aragon.org/{dao}">
```
Note that the urls need a `#/` before the DAO; e.g. `https://old-app.aragon.org/#/{dao}`.

Also, should this link have `target="_blank"`?

izqui replied:
Also this URL should probably be moved to a constant or a setting somewhere else, as we don't know yet what the URL will be :)

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+    body={<DeprecatedBody dao={dao} />}
+    onHide={onHide}
+    More={
+      <Button.Anchor mode="strong" href="URL" target="_blank">
```
Instead of the `'URL'`, we could also replace instances of `https://blog.aragon.one/0-5-dao-deprecation/` with a constant, especially as we don't know what the exact URL is yet.

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+          onClick={this.handleClick}
+          buttonText="More info"
+          color={negative}
+          dao={dao}
```
Do we need to pass `dao` as a prop here?

AquiGorka replied:
That's the way I was able to get it all the way down [here](https://github.com/aragon/aragon/pull/371/files/4fd00eedf9d5809fa84e27f40b477e88a155f994#diff-e341e1f76a6ca8ec7fe005cd3ac11853R62)

sohkai replied:
But I think `DeprecatedBody` gets rendered with the modal props, which is already handled [here](https://github.com/aragon/aragon/pull/371/files#diff-e341e1f76a6ca8ec7fe005cd3ac11853R24)?

The `Banner` doesn't seem to need the `dao` prop :).

AquiGorka replied:
My bad. Thanks for following through.

sohkai commented on src/components/Modal/Modal.js:
```diff
@@ -0,0 +1,66 @@
+import { Button, Text, breakpoint } from '@aragon/ui'
+
+const Modal = ({ title, body, moreText, onHide, More, blocking }) => (
+  <WrapModal role="alertdialog" onEscapeOutside={!blocking && onHide}>
```
⚠️ `onEscapeOutside={!blocking && onHide}` can be two types (`boolean` or `func`); I'd propose just making it a `func` by using a ternary: `blocking ? noop : onHide`

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+    return (
+      <React.Fragment>
+        <Banner
+          text="This Rinkeby DAO will be deprecated on 1/11/18"
```
I'd change this text to be `Old 0.5 Rinkeby DAOs will be deprecated on Nov. 1, 2018` since this appears during onboarding as well:

![image](https://user-images.githubusercontent.com/4166642/46488745-e0656a80-c803-11e8-97a4-7fb4d064f41b.png)

sohkai commented on src/components/ModalManager/ModalManager.js:
```diff
@@ -0,0 +1,92 @@
+        native
+        from={{ opacity: 0, top: 50 }}
+        enter={{ opacity: 1, top: 0 }}
+        leave={{ opacity: 0, top: 0 }}
```
Something that's not nice when doing this leave animation is that the user's blocked from being able to interact with the screen until the transition's completely done. I think @bpierre has encountered this before and solved it (maybe with the side panels?)... but not sure.

bpierre replied:
[`pointer-events: none`](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) 😆

AquiGorka replied:
`pointer-events` AND `hack-tricks` from @bpierre with `react-spring`

sohkai replied:
Can confirm this is working! Awesome guys! 💫

sohkai commented on src/Wrapper.js:
```diff
@@ -166,10 +172,14 @@ class Wrapper extends React.Component {
-      <React.Fragment>
+      <ModalProvider>
+        <ModalView />
+        {isDaoDeprecated && <DeprecatedBanner dao={dao} />}
```
I think the abstraction for the "notice banner" might not be at the right level. Given that we might reuse this "notice banner" later on, and have messages other than this deprecation notice, it might make sense to use a generic banner container here that we then render a body onto, rather than encapsulating it all into `DeprecatedBanner` as we are now.

Maybe components like `NoticeBannerContainer` and `GenericNoticeBannerBody` (that `DeprecatedBannerBody` would use) would fit?

AquiGorka replied:
Originally I added a `BannerView` (similar to the `ModalView` placeholder), yet there is no specific use-case for it so with @bpierre we thought it best to hardcode `DeprecatedBanner` for the time being, until we have the requirement to add another banner.

sohkai replied:
Sounds good! Let's refactor this later if/when the need arises.

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,94 @@
+      >
+        https://blog.aragon.one/0-5-dao-deprecation
+      </StyledSafeLink>{' '}
+      for more details.
```
I wrote this copy a while back... so some edits for myself :smile:

- Unfortunately **,** (add a comma after "Unfortunately)
- with the ~~currently deployed~~ contracts ~~that you might have been using~~ **your DAOs were constructed with** on Rinkeby.
- ~~So we've had to make~~ **As a result, we've made** the ~~hard~~ **tough** decision to deprecate ~~those~~ **all** old testnet DAOs.

sohkai commented on src/components/EscapeOutside/EscapeOutside.js:
```diff
@@ -0,0 +1,41 @@
+  }
+}
+
+export default EscapeOutside
```
Actually I just realized we use https://github.com/boblauer/react-onclickout in aragon-ui. Maybe we should just use that?

AquiGorka replied:
`react-onclickout` lacks the functionality to hide the container on an `esc` press
I could integrate it into `EscapeOutside`.
I'd like to hear your thoughts on this (@sohkai and @bpierre):

1. Integrate `onclickout` into `EscapeOutside` or,
1. Leave `EscapeOutside` as is, add it to `aragon-ui` and remove usage of `react-onclickout`

I like removing dependencies when possible, but in this case, it means adding task(s) to the backlog for _who-knows-when_...

bpierre replied:
I think it could be nice to have an `Overlay` component in Aragon UI, having:

- A background color + size + positioning.
- The opacity transition (by passing a react-spring transition object so we can sync it with another transition).
- An `onClick` event.

The `esc` event would be managed by the components using the Overlay, like the Dialog or the SidePanel.

About using onclickout: we probably don’t need any of react-onclickout or a custom solution, since we control everything that is outside of the component (i.e. the Overlay) and the entire view is blocked. `onClick` should be enough?

sohkai replied:
:+1: we could also add the `pointer-events: none` to `Overlay` so we don't run into this problem again with overlays :).

Something nice about `react-onclickout` is that it includes some support for touch events, and that would be nice to add into our `Overlay`.

AquiGorka commented on src/components/Banner/Banner.js:
```diff
@@ -0,0 +1,28 @@
+  grid-template-columns: auto auto;
+  grid-gap: 1em;
+  align-items: center;
+  justify-content: center;
```
Yes, they are. If I remove `align-items` the content will not align vertically to the center (top) and if I remove `justify-content` both items will use all the available space and the text becomes left aligned.

sohkai replied:
Nevermind me, I thought those two were only for `display: flexbox` 🙈

AquiGorka commented on src/components/Modal/Modal.js:
```diff
@@ -0,0 +1,78 @@
+    padding: 3.5em;
+    margin: auto;
+  `
+  )} ${breakpoint(
```
Yeah, I agree with your formatting preference but `prettier` formatted the code this way. I tried to add the extra line break and indentation on the backtick and the `pre-push` hook would not allow me to submit the changes... advice on how to proceed?

AquiGorka replied:
Thanks @bpierre for letting me know how to make this work.

AquiGorka commented on src/components/Modal/Modal.js:
```diff
@@ -0,0 +1,78 @@
+import React from 'react'
```
I thought we could add Modal to `aragon-ui` (eventually), that's why I added it in its own folder, what do you think?

AquiGorka commented on src/components/ModalManager/ModalManager.js:
```diff
@@ -0,0 +1,91 @@
+    return (
+      <ModalContext.Provider value={this.state}>
+        {this.props.children}
+      </ModalContext.Provider>
```
I really liked this change, although I was not able to remove the `z-index` attribute but did change it to `2`.

sohkai replied:
Yeah, just realized the stacking contexts we have for the main app (not in onboarding) are a bit messy... Let's fix that at a later time :).

AquiGorka commented on src/components/ModalManager/ModalManager.js:
```diff
@@ -50,42 +37,56 @@ class ModalProvider extends Component {
+  ${breakpoint(
+    'medium',
+    `
+    height: unset;
```
Good catch! I was able to remove the breakpoint by setting `Wrap`'s display to `grid`

sohkai replied:
I was playing around with this in the inspector and with just

```js
const AnimatedWrap = styled(animated.div)`
  position: relative;
`
```

I think both small and larger displays are aligning to the center.

sohkai commented on src/components/EscapeOutside/EscapeOutside.js:
```diff
@@ -14,15 +14,15 @@ class EscapeOutside extends React.Component {
   handleClick = e => {
-    const { onEscapeOutside } = this.props
-    if (!this.node.contains(e.target) && onEscapeOutside) {
+    const { onEscapeOutside = () => {} } = this.props
```
A default prop is better than a default assignment, as it avoids recreating the `noop` on each method invocation.

sohkai commented:
Looking really good @AquiGorka, thanks for making the changes :)

A few leftover collections of comments (some of which are hidden):

- https://github.com/aragon/aragon/pull/371#discussion_r223064369
- https://github.com/aragon/aragon/pull/371#discussion_r222917055 cc @bpierre 
- https://github.com/aragon/aragon/pull/371#discussion_r223050030
- https://github.com/aragon/aragon/pull/371#discussion_r223071576

Given the added complexity of having a self-contained [`Overlay` component](https://github.com/aragon/aragon/pull/371#discussion_r223057680), @bpierre what do you think about working on that in aragon-ui separately (and also removing `react-onclick`) and updating aragon/aragon afterwards so we can roll this out sooner?

AquiGorka commented:
@bpierre and @sohkai I've addressed all the comments. My last [commit](https://github.com/aragon/aragon/pull/371/commits/726ead9a551fc8a6876d778aa26d0fb833d28061) moves `ModalProvider` and `DeprecatedBanner` up to `App` (easier to revert in case we want to keep it at `Wrapper` level).

Let me know if there's anything else.

sohkai commented:
> It is also the first time that we use the term “DAO” instead of “organization” (in the dialog too). I think we should consider replacing “DAO” by “organization”, or at the minimum use them both to help users understand what we are referring to.

Good point, we should just use `organization`.

sohkai commented:
@izqui Yes, see the second point in https://github.com/aragon/design/issues/12, which was mostly implemented in #369.

sohkai commented on src/components/Banner/Banner.js:
```diff
@@ -1,27 +1,35 @@
+    border: '0',
+  }
+  render() {
+    const { text, textColor, button, buttonText, onClick, color } = this.props
```
Seems like we should just remove `buttonText` and `onClick` at this point.

sohkai commented on src/components/ModalManager/ModalManager.js:
```diff
@@ -0,0 +1,105 @@
+
+const AnimatedWrap = styled(animated.div)`
+  position: relative;
+  min-height: 0;
```
Just wondering, was this ever set to something negative? Or does flag the layout engine to do something?

sohkai commented on src/components/ModalManager/Modal.js:
```diff
@@ -0,0 +1,65 @@
+    `
+  )};
+
+  ${breakpoint('large', `max-width: 70%;`)};
```
What do you think about adding a `max-width: 800px` to all breakpoints? On larger displays this can look quite awkward:

<img width="1171" alt="screen shot 2018-10-12 at 11 08 58 am" src="https://user-images.githubusercontent.com/4166642/46859880-4eff8500-ce0f-11e8-8b0a-36ddfeccc8cd.png">

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -0,0 +1,108 @@
+const StyledSafeLink = styled(SafeLink)`
+  text-decoration-color: ${accent};
+  color: ${accent};
+  word-break: break-all;
```
@bpierre I think this should be `break-word`, so that the links (which aren't that long) don't get awkwardly line split.

bpierre replied:
Agree, we should always avoid `break-all`. 👍

sohkai commented on src/components/DeprecatedBanner/DeprecatedBanner.js:
```diff
@@ -64,28 +64,27 @@ const DeprecatedBody = ({ dao }) => (
-      no longer be accessible on app.aragon.one. To help migrate, and in case
-      you still need access to this organization, we will continue hosting this
-      version of Aragon Core on{' '}
-      <StyledSafeLink href={`${OLD_DAO_URL}${dao}`} target="_blank">
```
@bpierre I think this old behaviour was still nice; it would try to open the same DAO on the old app instance in case it was available there. Now we'd just bring them to a (dead) onboarding screen.

bpierre commented on src/components/ModalManager/Modal.js:
```diff
@@ -0,0 +1,65 @@
+      </Header>
+      <Body>{body}</Body>
+      <Footer>
+        <Button mode="text" onClick={onHide}>
```
Yes, we should probably have a prop for text buttons to make the right / left padding optional for these cases.

bpierre commented:
@AquiGorka @sohkai shipping ship party 🌊🌊🚢🎊🕺💃🎉🚢🌊🌊

AquiGorka commented:
Yay!! Thanks @bpierre and @sohkai !
